src/librustc/mir/interpret/error.rs

```diff
@@ -342,27 +342,33 @@ impl fmt::Debug for InvalidProgramInfo<'tcx> {
 
 #[derive(Clone, RustcEncodable, RustcDecodable, HashStable)]
 pub enum UndefinedBehaviorInfo {
-    /// Handle cases which for which we do not have a fixed variant.
+    /// Free-form case. Only for errors that are never caught!
     Ub(String),
+    /// Free-form case for experimental UB. Only for errors that are never caught!
+    UbExperimental(String),
     /// Unreachable code was executed.
     Unreachable,
+    /// An `assume` was run on a `false` condition,
+    AssumptionNotHeld,
```

Comment:
Is this ever caught? If not, remove the variant

Reply:
It is not caught, and neither is `Unreachable`...

I don't have a clear handle yet on which errors should get their own variants. What I *don't* like is things like the `Function*Mismatch` family of variants: those errors have a lot of context-sensitive information in them and are never caught, so we can just create the string locally and include all the information we want. The noise of having to pack this data into an enum and then unpack it later is just annoying.

In contrast, `Unreachable` and `AssumptionNotHeld` are "classical" cases of UB that don't need any extra context (there is no information to give besides "this bad thing happened"), so I am not very offended by having them as separate variants. Though `AssumptionViolated` would be a better name.

Reply:
But on the other hand it is true that if `AssumptionNotHeld` didn't exist yet, I would certainly not add it when implementing that intrinsic in Miri. And `unreachable` is at least thrown in several places, so it serves to dedup the message.

So I will remove `AssumptionNotHeld`.